src/core/cycle_detector.go

```diff
@@ -5,108 +5,60 @@ import (
-	deps  map[*BuildLabel][]*BuildLabel
-	queue chan dependencyLink
-}
-
-func newCycleDetector() *cycleDetector {
-	c := new(cycleDetector)
-
-	// Set this to an order of magnitude that seems sensible to avoid resizing them rapidly initially
-	c.deps = make(map[*BuildLabel][]*BuildLabel, 1000)
-
-	// Buffer this a little so handle spikes in targets coming in
-	c.queue = make(chan dependencyLink, 1000)
-
-	return c
+	graph *BuildGraph
 }
 
-// AddDependency queues up another dependency for the cycle detector to check
-func (c *cycleDetector) AddDependency(from *BuildLabel, to *BuildLabel) {
-	go func() {
-		c.queue <- dependencyLink{from: from, to: to}
-	}()
-}
-
-// checkForCycle just checks to see if there's a dependency cycle. It doesn't compute the cycle to avoid excess
-// allocations. buildCycle can be used to reconstruct the cycle once one has been found.
-func (c *cycleDetector) checkForCycle(head, tail *BuildLabel, done map[BuildLabel]struct{}) bool {
-	if _, ok := done[*tail]; ok {
-		return false
-	}
-	done[*tail] = struct{}{}
-	if tailDeps, ok := c.deps[tail]; ok {
-		for _, dep := range tailDeps {
-			if dep == head {
-				// If the tail has a dependency on the head, we've found a cycle
-				return true
-			}
-
-			if c.checkForCycle(head, dep, done) {
-				return true
+// Check runs a single check of the build graph to see if any cycles can be detected.
+// If it finds one an errCycle is returned.
+func (c *cycleDetector) Check() *errCycle {
+	log.Debug("Running cycle detection...")
+	permanent := map[*BuildTarget]struct{}{}
+	temporary := map[*BuildTarget]struct{}{}
+
+	var visit func(target *BuildTarget) ([]*BuildTarget, bool)
+	visit = func(target *BuildTarget) ([]*BuildTarget, bool) {
+		if _, present := permanent[target]; present {
+			return nil, false
+		} else if _, present := temporary[target]; present {
+			return []*BuildTarget{target}, false
+		}
+		temporary[target] = struct{}{}
+		for _, dep := range target.Dependencies() {
+			if cycle, done := visit(dep); cycle != nil {
+				if done || target == cycle[len(cycle)-1] {
+					return cycle, true // This target is already in the cycle
+				}
+				return append([]*BuildTarget{target}, cycle...), false
 			}
 		}
+		delete(temporary, target)
+		permanent[target] = struct{}{}
+		return nil, false
 	}
-	return false
-}
-
-// buildCycle is used to actually reconstruct the cycle after we've found one
-func (c *cycleDetector) buildCycle(chain []*BuildLabel, done map[BuildLabel]struct{}) []*BuildLabel {
-	tail := chain[len(chain)-1]
-	head := chain[0]
-
-	if _, ok := done[*tail]; ok {
-		return nil
-	}
-	done[*tail] = struct{}{}
-	if tailDeps, ok := c.deps[tail]; ok {
-		for _, dep := range tailDeps {
-			if dep == head {
-				// If the tail has a dependency on the head, we've found a cycle
-				return append(chain, dep)
-			}
 
-			if newChain := c.buildCycle(append(chain, dep), done); newChain != nil {
-				return newChain
+	for _, target := range c.graph.AllTargets() {
```

Comment:
I wonder if we can get the queued targets that are waiting on their deps somehow?